spec/fixtures/generate_external_certs.sh

```diff
@@ -0,0 +1,337 @@
+  
+  mkdir -p /tmp/ca_gen/keys
+  DIR_KEYS=/tmp/ca_gen/keys
+  
+  mkdir -p /tmp/ca_gen/cadata
+  DIR_CADATA=/tmp/ca_gen/cadata
+  
+  echo "dir cadata: $DIR_CADATA"
+}
+
+function cleanup() {
+  rm -rf /tmp/ca_gen
+}
+
+function init_ca() { # $1 ca_name
+  echo "Initiating CA  \"$1\""
+  generate_conf $1
+  # Create / clear certs directory for CA
+  mkdir -p $DIR_CADATA/$1/db.certs
+  rm $DIR_CADATA/$1/db.certs/* 2>/dev/null
+  
+  # Create / clear index file of certs
+  rm $DIR_CADATA/$1/db.inde* 2>/dev/null
+  touch $DIR_CADATA/$1/db.index
+  
+  # Other basic needed files
+  echo unique_subject = yes > $DIR_CADATA/$1/db.index.attr
+  echo 01 > $DIR_CADATA/$1/db.serial
+  
+  # Create the keyfile for the CA
+  openssl genrsa -out $DIR_KEYS/ca_$1.key 2>/dev/null
+}
+
+function init_root_ca() { # $1 ca_name, $2 subject
+  init_ca $1
+  echo "Generating CRT \"$1\""
+  openssl req -new -x509 -days 100 -extensions v3_ca -key $DIR_KEYS/ca_$1.key -out $DIR_CERTS/ca_$1.crt -subj "$2"
+}
+
+function init_sub_ca() { # $1 ca_name, $2 parent_ca, $3 subject
+  init_ca $1
+  echo "Generating CRT \"$1\""
+  openssl req -sha256 -new -key $DIR_KEYS/ca_$1.key -out $CSR_DIR/ca_$1.csr -subj "$3"
+  openssl ca -batch \
+    -config $DIR_CONF/ca_$2.conf \
+    -extensions v3_ca \
+    -out $DIR_CERTS/ca_$1.crt \
+    -infiles $CSR_DIR/ca_$1.csr 2>/dev/null
+}
+
+function append_altnames() { # $1 altnames, $2 filename
+  IFS='-' read -r -a altnames <<< "$1"
+  for index in "${!altnames[@]}"
+  do
+      echo "DNS.$(($index+1)) = ${altnames[index]}" >> $2
+  done
+}
+
+function append_ip_altnames() { # $1 altnames, $2 filename
+  IFS='-' read -r -a altnames <<< "$1"
+  for index in "${!altnames[@]}"
+  do
+      echo "IP.$(($index+1)) = ${altnames[index]}" >> $2
+  done
+}
+
+function site_request_conf() { # $1 file, $2 bits
+  echo "[req]" > $1
+  echo default_bits = $2 >> $1
+  echo distinguished_name = req_distinguished_name >> $1
+  echo req_extensions = req_ext >> $1
+  
+  # Various prompts for the portions of the subject would go here, as well as defaults.
+  # They can, though, be left blank when the subject is provided on the command line.
+  cat >>$1 <<EOL
+[req_distinguished_name]
+EOL
+  
+  echo -e "[req_ext]\nsubjectAltName = @alt_names\n" >> $1
+  echo -e "[alt_names]" >> $1
+}
+
+function gen_start_end() { # $1 good/expired
+  # Right now we are just using harded good/bad for the date range validity.
+  # For most purposes it would be better to expose the clean text specification
+  # of both start and end as parameters.
+  if [ $1 = "good" ]
+  then
+    START_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "-1 day")
+    END_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "+5 years")
+  else
+    START_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "-10 day")
+    END_DATE=$(TZ=UTC date +"%Y%m%d%H%M%SZ" -d "-5 days")
+  fi
+}
+
+function create_site_crt() { # $1 domain, $2 ca, $3 message_digest, $4 rsa_bits, $5 good/expired, $6 subject, $7 altnames, $8 ipaltnames
```

Comment:
There are a lot of parameters here. Named parameters could potentially be used by way of getopts, but has been avoided here to make it plain bash without requiring getopts. Additionally using getopts for function calls within a bash script is somewhat non-standard and weird.

Reply:
It may be reasonable to use more environment variables for the arguments so that the code is more clear to read. I dislike polluting environment variable space so I have not done so here.

Reply:
getopts is implemented as a builtin in bash, so there's really no extra requirement...  I agree that using global env vars is the worst option, though.

I prefer to document this in my scripts by assigning the positional parameters to descriptive variable names, which makes the rest of the function a tad easier to read.
```bash
function fname {
  something=$1
  anotherthing=$2 # another thing
}
```

Reply:
Have changed to using named parameters using associative arrays thanks to your help in chat on how to do that. Thanks for all the ideas; it enhances the readability of the script a bunch.